### PR TITLE
SWATCH-2650 Implement level 1 and level 2 attr in offering

### PR DESCRIPTION
**Pull request by kartikshahc**
Jira issue: SWATCH-2650

## Description

We'll be associating Ansible skus with swatch product tags via their the Level 1 and Level 2 attributes on the sku.  A swatch product definition matches when BOTH level 1 and level 2 match.

- [ ] `offering` database record should persist Level1 and Level2 values after an offering sync
- [ ] `sku_product_tag` database records should consider Level1 and Level2 values
- [ ] Events only match to `ansible-aap-managed` doesn't get "detected" on any events...The only way we'd persist an event with `ansible-aap-managed` is if it was explicitly set in the Event message

## Setup

```
PRODUCT_URL=http://localhost:8101/svcrest/product/v3 DEV_MODE=true ./gradlew :bootRun
```

```
CONTRACT_USE_STUB=true QUARKUS_MANAGEMENT_PORT=9002 SERVER_PORT=8002 DEV_MODE=true ./gradlew :swatch-billable-usage:quarkusDev
```

## Testing

Made a pytest project for functional/isolated tests for this that can be found on [a branch in our swatch-support-scripts repo](https://gitlab.cee.redhat.com/rhsm/swatch-support-scripts/-/merge_requests/10/diffs).  The gist is

- Sync offering for sku MCT4701.  My tests use Wiremock for this.  If you don't want to mock it out, you'll need to point PRODUCT_URL to production because MCT4701 isn't live in stage yet.  See https://redhat-internal.slack.com/archives/C01F7QFNATC/p1720542569145639?thread_ts=1720534101.684429&cid=C01F7QFNATC for more details

```bash
http PUT :8000/api/rhsm-subscriptions/v1/internal/rpc/offerings/sync/MCT4701 x-rh-swatch-psk:placeholder | jq
```

If your database doesn't already have it
```json
{
  "detail": "syncResult=SKIPPED_MATCHING (Upstream matches stored item, did not sync) for offeringSku=\"MCT4701\"."
}
```
If your database already has it:
```json
{
    "detail": "syncResult=SKIPPED_MATCHING (Upstream matches stored item, did not sync) for offeringSku=\"MCT4701\"."
}
```

- Check the database to make sure MCT4701 shows up with the appropriate tag
```sql
select * from sku_product_tag where sku = 'MCT4701';
```
| sku | product\_tag |
| :--- | :--- |
| MCT4701 | ansible-aap-managed |

- Check the offering/product_tag API 
```bash
http GET :8000/api/rhsm-subscriptions/v1/internal/offerings/MCT4701/product_tags x-rh-swatch-psk:placeholder | jq
```
```json
{
  "data": [
    "ansible-aap-managed"
  ]
}
```

### CODE REVIEW

File: src/main/java/org/candlepin/subscriptions/product/UpstreamProductData.java [hunk 1]
```diff
@@ -63,6 +63,8 @@ private enum Attr {
     IFL,
     SOCKET_LIMIT,
     SERVICE_TYPE,
+    LEVEL_1,
```
Comment (kartikshahc) [inline src/main/java/org/candlepin/subscriptions/product/UpstreamProductData.java:66]: Not sure if we need '_'  but, to be consistent may be we can add it to offering table columns?

File: swatch-product-configuration/src/main/java/com/redhat/swatch/configuration/registry/SubscriptionDefinition.java [hunk 2]
```diff
@@ -385,6 +387,28 @@ private static Predicate<Variant> createMetricIdsPredicate(ProductTagLookupParam
...
+              && Objects.equals(paramLevel1, variantLevel1);
+        };
+
+    Predicate<Variant> level2Predicate =
+        variant -> {
+          String paramLevel2 = params.getLevel2();
+          String variantLevel2 = variant.getLevel2();
+          return !isNullOrEmpty(paramLevel2)
+              && !isNullOrEmpty(variantLevel2)
+              && Objects.equals(paramLevel2, variantLevel2);
```
Comment (kahowell) [inline swatch-product-configuration/src/main/java/com/redhat/swatch/configuration/registry/SubscriptionDefinition.java:406]:
```suggestion
          return isNullOrEmpty(paramLevel2)
              || (!isNullOrEmpty(variantLevel2)
              && Objects.equals(paramLevel2, variantLevel2));
```

**Review by kahowell — CHANGES_REQUESTED** [review]

**Review by kahowell — APPROVED** [review]
All is good, I misunderstood the implementation previously, my bad :smile:

**Review by kahowell — APPROVED** [review]

**Review by liwalker-rh — APPROVED** [review]
Testing looks good. Also working on IQE coverage: https://gitlab.cee.redhat.com/insights-qe/iqe-rhsm-subscriptions-plugin/-/merge_requests/791

**Pull request merged**